File: libs/bot-engine/src/lib/move-encoder.ts
/**
 * Encodes/decodes UCI chess moves to/from a flat policy index (0–4095).
 *
 * Index = fromSquare * 64 + toSquare
 * where squares are numbered 0–63: a1=0, b1=1, ..., h8=63
 *
 * Promotions: queen promotion is assumed when the last rank is reached.
 * Other promotions (n/b/r) are encoded as the base from-to pair
 * and resolved externally if needed.
 */

function squareIndex(squareStr: string): number {
  const file = squareStr.charCodeAt(0) - 'a'.charCodeAt(0);
  const rank = parseInt(squareStr[1], 10) - 1;
  return rank * 8 + file;
}

function squareFromIndex(idx: number): string {
  const file = idx % 8;
  const rank = Math.floor(idx / 8);
  return String.fromCharCode('a'.charCodeAt(0) + file) + String(rank + 1);
}

export function uciToIndex(uci: string): number {
  const from = squareIndex(uci.slice(0, 2));
  const to = squareIndex(uci.slice(2, 4));
  return from * 64 + to;
}

export function indexToUci(index: number): string {
  const from = Math.floor(index / 64);
  const to = index % 64;
  return squareFromIndex(from) + squareFromIndex(to);
}

/**
 * Builds a Float32Array mask of length 4096.
 * Legal moves get value 1, all others 0.
 */
export function buildLegalMask(legalMoves: string[]): Float32Array {
  const mask = new Float32Array(4096);
  for (const move of legalMoves) {
    const idx = uciToIndex(move);
    if (idx >= 0 && idx < 4096) {
      mask[idx] = 1;
    }
  }
  return mask;
}

File: chess-bot-trainer/src/encoder/move_encoder.py
"""
Encodes/decodes UCI moves to/from a flat policy index 0–4095.
Must stay in sync with libs/bot-engine/src/lib/move-encoder.ts.

Index = from_square * 64 + to_square
Squares: a1=0, b1=1, ..., h8=63
"""

import chess
import numpy as np


def uci_to_index(uci: str) -> int:
    move = chess.Move.from_uci(uci)
    return move.from_square * 64 + move.to_square


def index_to_uci(index: int) -> str:
    from_sq = index // 64
    to_sq = index % 64
    return chess.Move(from_sq, to_sq).uci()


def build_policy_target(uci: str) -> np.ndarray:
    """One-hot vector of shape (4096,)."""
    target = np.zeros(4096, dtype=np.float32)
    idx = uci_to_index(uci)
    target[idx] = 1.0
    return target


def build_legal_mask(board: chess.Board) -> np.ndarray:
    """Float32 mask of shape (4096,): 1 for legal moves, 0 otherwise."""
    mask = np.zeros(4096, dtype=np.float32)
    for move in board.legal_moves:
        idx = move.from_square * 64 + move.to_square
        mask[idx] = 1.0
    return mask
